### PR TITLE
[macOS] Added CodeQL Action Bundle to macOS 11 and macOS 12

### PR DESCRIPTION
**Pull request by erik-bershel**
# Description
Added CodeQL Action bundle to cached tools. Only partly tested.

#### Related issue:
Previous PR issue #6059

## Virtual environments affected
- [ ] macOS 10.15
- [x] macOS 11
- [x] macOS 12

## Check list
- [x] Related issue / work item is attached
- [x] Tests are written (if applicable)
- [x] Documentation is updated (if applicable)
- [ ] Changes are tested and related VM images are successfully generated

### CODE REVIEW

File: images/macos/toolsets/toolset-11.json [hunk 1]
```diff
@@ -296,6 +296,14 @@
                 "3.0.*",
                 "3.1.*"
             ]
+        },
+        {
+            "name": "CodeQL",
```
Comment (al-cheb) [inline images/macos/toolsets/toolset-11.json:301]: Where do we use it?

File: images/macos/software-report/SoftwareReport.Toolcache.psm1 [hunk 2]
```diff
@@ -57,6 +65,8 @@ function Build-ToolcacheSection {
     $output += New-MDList -Lines (Get-ToolcacheNodeVersions) -Style Unordered
     $output += New-MDHeader "Go" -Level 4
     $output += Get-ToolcacheGoTable
+    $output += New-MDHeader "CodeQL" -Level 4
```
Comment (al-cheb) [inline images/macos/software-report/SoftwareReport.Toolcache.psm1:68]: Let's for consistency use the output like for Windows and Linux.

File: images/macos/provision/core/codeql-bundle [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash -e -o pipefail
```
Comment (al-cheb) [inline images/macos/provision/core/codeql-bundle:1]: Where do you update packer templates to add this script?

Comment (al-cheb) [inline images/macos/provision/core/codeql-bundle:1]: Let's for consistency add the .sh extension

File: images/macos/toolsets/toolset-11.json [hunk 4]
```diff
@@ -296,6 +296,14 @@
...
+            "arch": "x64",
+            "platform" : "darwin",
+            "versions": [
+                "*"
```
Comment (al-cheb) [inline images/macos/toolsets/toolset-11.json:305]: How do we plan to install multiple versions?

File: images/macos/software-report/SoftwareReport.Toolcache.psm1 [hunk 5]
```diff
@@ -44,6 +44,14 @@ function Get-ToolcacheGoTable {
     return $Content
 }
 
+function Get-CodeQLBundleVersion {
```
Comment (al-cheb) [inline images/macos/software-report/SoftwareReport.Toolcache.psm1:47]: If we have an array of versions why do we return only one version?

File: images/macos/toolsets/toolset-12.json [hunk 6]
```diff
@@ -241,6 +241,14 @@
                 "3.0.*",
                 "3.1.*"
             ]
+        },
+        {
+            "name": "CodeQL",
+            "arch": "x64",
+            "platform" : "darwin",
```
Comment (al-cheb) [inline images/macos/toolsets/toolset-12.json:248]: Where do we use this field?

File: images/macos/software-report/SoftwareReport.Toolcache.psm1 [hunk 7]
```diff
@@ -44,6 +44,14 @@ function Get-ToolcacheGoTable {
...
+    $CodeQLVersionsWildcard = Join-Path $Env:AGENT_TOOLSDIRECTORY -ChildPath "CodeQL" | Join-Path -ChildPath "*"
+    $CodeQLVersionPath = Get-ChildItem $CodeQLVersionsWildcard | Select-Object -First 1 -Expand FullName
+    $CodeQLPath = Join-Path $CodeQLVersionPath -ChildPath "x64" | Join-Path -ChildPath "codeql" | Join-Path -ChildPath "codeql"
+    $CodeQLVersion = & $CodeQLPath version --quiet
+    return "CodeQL Action Bundle $CodeQLVersion"
```
Comment (al-cheb) [inline images/macos/software-report/SoftwareReport.Toolcache.psm1:52]: If we are planning to use multiple version it should be like:

CodeQL Action Bundle:
- 0.0.0.1
- 0.0.0.2
- etc

**Comment by erik-bershel** [comment]
@al-cheb changes:
1) Added extension to shell-script file codeql-bundle
2) Rewrote Software Reporter section
3) Rewrote test to simplier version without versions control
4) Added changes to provision scripts

File: images/macos/provision/core/codeql-bundle.sh [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+codeql_bundle_version="0.0.0-${codeql_bundle_name##*-}"
+
+echo "Downloading CodeQL bundle $codeql_bundle_version..."
+download_with_retries "https://github.com/github/codeql-action/releases/download/$codeql_bundle_name/codeql-bundle.tar.gz" "/tmp" "codeql-bundle.tar.gz"
+codeqlArchive="/tmp/codeql-bundle.tar.gz"
+
+codeqlToolcachePath="$AGENT_TOOLSDIRECTORY/codeql/$codeql_bundle_version/x64"
+mkdir -p $codeqlToolcachePath
+
+echo "Unzip CodeQL archive..."
```
Comment (mikhailkoliada) [inline images/macos/provision/core/codeql-bundle.sh:16]: forthcoming command is not unzip

**Review by Butterbean8 — APPROVED** [review]

**Review by Butterbean8 — APPROVED** [review]

File: images/macos/provision/core/codeql-bundle.sh (hunk 8 above)
Comment (erik-bershel) [reply images/macos/provision/core/codeql-bundle.sh]: Corrected the wording to be closer to the truth.

File: images/macos/toolsets/toolset-11.json (hunk 1 above)
Comment (erik-bershel) [reply images/macos/toolsets/toolset-11.json]: Removed this part as recommended.

**Review by ddobranic — APPROVED** [review]

**Pull request merged**